ZGrab2 Framework

ajholland opened the pull request:
DO NOT MERGE

This is my proposed framework for zgrab2. This is by no means complete and is missing parts along with containing some rough edges (commented out code).

The main goals of this are to ensure a sane commandline, ability to run multiple scans at once, and ability to easily add new protocols. Below are examples of this:
`./zgrab2 --help`
`./zgrab2 http --help`
`./zgrab2 http --endpoint="/.well-known"`
`./zgrab2 tls --user-agent="bob"`
`./zgrab2 -o zgrabOutput http -p 70` 
`./zgrab2 mult -c mult.ini`

Defaults are set on structs as well as short and long names. Short names (one character) are only for upper level options. The main idea is that the same struct that is used for the options of a protocol will also contain the fields for results of the protocol. This struct is marshaled and output at the end. This achieves the goal of easily adding new protocols. By specifying short names/long names it becomes an option, otherwise it is a field that will be added by the protocol and output at the end. 

At present this program accepts no input but pretends as if 2 IP's have been entered into a process queue.

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	OutputFileName     string     `short:"o" long:"output-file" default:"-" description:"Output filename, use - for stdout" json:"out"`
+	InputFileName      string     `short:"i" long:"input-file" default:"-" description:"Input filename, use - for stdin" json:"in"`
+	Meta               string     `short:"m" long:"metadata-file" default:"-" description:"Metadata filename, use - for stdout" json:"meta"`
+	Log                string     `short:"l" long:"log-file" default:"-" description:"Log filename, use - for stdout" json:"log"`
```
This should probably be named consistently with OutputFileName and InputFileName as MetadataFileName. The same for Log.

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	InputFileName      string     `short:"i" long:"input-file" default:"-" description:"Input filename, use - for stdin" json:"in"`
+	Meta               string     `short:"m" long:"metadata-file" default:"-" description:"Metadata filename, use - for stdout" json:"meta"`
+	Log                string     `short:"l" long:"log-file" default:"-" description:"Log filename, use - for stdout" json:"log"`
+	LookupDomain       bool       `short:"d" long:"lookup-domain" description:"Input contains only domain names" json:"lookup"`
```
This doesn't really make sense as flag. Why can't we autodetect this?

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	Meta               string     `short:"m" long:"metadata-file" default:"-" description:"Metadata filename, use - for stdout" json:"meta"`
+	Log                string     `short:"l" long:"log-file" default:"-" description:"Log filename, use - for stdout" json:"log"`
+	LookupDomain       bool       `short:"d" long:"lookup-domain" description:"Input contains only domain names" json:"lookup"`
+	Interface          string     `short:"f" long:"interface" description:"Network interface to send on" json:"inter"`
```
This should absolutely be `-i` not `-f`.  🔌

ajholland replied:
-f should be input file

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	LookupDomain       bool       `short:"d" long:"lookup-domain" description:"Input contains only domain names" json:"lookup"`
+	Interface          string     `short:"f" long:"interface" description:"Network interface to send on" json:"inter"`
+	Timeout            int        `short:"t" long:"timeout" description:"Set connection timeout in seconds" json:"time"`
+	GOMAXPROCS         int        `short:"g" long:"gomaxprocs" default:"3" description:"Set GOMAXPROCS" json:"gomax"`
```
This should default to OS one not to a static number.

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	GOMAXPROCS         int        `short:"g" long:"gomaxprocs" default:"3" description:"Set GOMAXPROCS" json:"gomax"`
+	Senders            int        `short:"s" long:"senders" default:"1000" description:"Number of send coroutines to use" json:"sender"`
+	ConnectionsPerHost int        `short:"h" long:"connections-per-host" defaults:"1" description:"Number of times to connect to each host (results in more output)" json:"connect"`
+	Prometheus         string     `short:"a" long:"prometheus" description:"Address to use for Prometheus server (e.g. localhost:8080). If empty, Prometheus is disabled." json:"-"`
```
I probably wouldn't give this a short.

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+	Timeout            int        `short:"t" long:"timeout" description:"Set connection timeout in seconds" json:"time"`
+	GOMAXPROCS         int        `short:"g" long:"gomaxprocs" default:"3" description:"Set GOMAXPROCS" json:"gomax"`
+	Senders            int        `short:"s" long:"senders" default:"1000" description:"Number of send coroutines to use" json:"sender"`
+	ConnectionsPerHost int        `short:"h" long:"connections-per-host" defaults:"1" description:"Number of times to connect to each host (results in more output)" json:"connect"`
```
no short

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+			if err := foo.Parse(bar); err != nil {
+				log.Fatal(err)
+			}
+			switch scanner.Text()[:3] { //todo: remove this shit, fork goflags or recreate
```
This needs to be dynamic.

zakird commented on zlib/controller.go:
```diff
@@ -0,0 +1,53 @@
+package zlib
+
+// Controller implements worker interface
+type Controller struct {
```
Rename to something specific to metadata file. E.g., monitor

cdzombak replied:
Yep, this definitely needs a clearer name.

zakird commented on zlib/controller.go:
```diff
@@ -0,0 +1,53 @@
+
+// Controller implements worker interface
+type Controller struct {
+	success  uint
```
These should probably be plural since singular sounds like a boolean.

zakird commented on zlib/processing.go:
```diff
@@ -0,0 +1,102 @@
+				panic(err.Error())
+			}
+			if _, err := out.Write([]byte("\n")); err != nil {
+				panic(err.Error())
```
Fatal

zakird commented on zlib/processing.go:
```diff
@@ -0,0 +1,102 @@
+	go func() {
+		for result := range outputQueue {
+			if _, err := out.Write(result); err != nil {
+				panic(err.Error())
```
Fatal

zakird commented on zlib/processing.go:
```diff
@@ -0,0 +1,102 @@
+		go func() {
+			for obj := range processQueue {
+				//divide up, run, consolidate
+				result := GrabWorker(obj)
```
`GrabWorker` makes me think that you were grabbing a worker. You might want to name this something like `MakeZGrabWorker`

zakird replied:
Serialize/Linearize Grab orker.

There should only be --senders numbers of works and they should process things in a single thread.

zakird commented on mult.ini:
```diff
@@ -0,0 +1,10 @@
+[http]
```
allow specify continue on failure.

zakird commented on zlib/config.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// this is a super hacky method of ensuring that all options get defaults set
+func deepCopyAll() {
```
This will go away with new library.

zakird commented on zlib/processing.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// GrabWorker divides up input and sends to each handler and then consolidates at end
+func GrabWorker(input interface{}) []byte {
```
(rename function)

zakird commented on zlib/processing.go:
```diff
@@ -0,0 +1,102 @@
+
+	protocolResult := make(map[string]interface{})
+
+	for i := 0; i < NumProtocols; i++ {
```
Actions

zakird commented on zlib/ssh.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+// Execute validates the options sent to SSHConfig and then passes operation back to main
+func (x *SSHConfig) Execute(args []string) error {
```
name this better

zakird commented on zlib/tls.go:
```diff
@@ -0,0 +1,40 @@
+package zlib
+
```
There needs to be

(1) a specific flag validation call
(2) a per-module init call
(3) a per-module per-routine init call

cdzombak commented:
It might be worthwhile to update the `README` on this branch with a quick overview of how configuration works. Writing documentation as we go should help us avoid an outcome like the current state of zgrab1 documentation (https://github.com/zmap/zgrab/issues/132, https://github.com/zmap/zgrab/issues/129, https://github.com/zmap/zgrab/issues/78, https://github.com/zmap/zgrab/issues/31, https://github.com/zmap/zgrab/issues/28, https://github.com/zmap/zgrab/issues/3).

> The main idea is that the same struct that is used for the options of a protocol will also contain the fields for results of the protocol. This struct is marshaled and output at the end. This achieves the goal of easily adding new protocols. By specifying short names/long names it becomes an option, otherwise it is a field that will be added by the protocol and output at the end.

Without having read the code, I don't see what reusing this struct gives us. My mental model, at a high level, is something like: "Options is parsed into a struct, which we treat as immutable once the options have been parsed. That struct is used as input for the grabber, which produces some JSON-serializable structs as output."

Why doesn't that work? How does reusing this one struct "achieve the goal of easily adding new protocols"?

cdzombak commented:
Looking at the organization of this project — again, before digging into the code — it seems like you have protocol definitions (`tls.go`, `ssh.go` etc) under `zlib` alongside `config.go` and `processing.go`. Especially when considering that we may want to add many, many more protocols, it would make sense to better organize this project so the zgrab core is clearly separated from protocol-specific files.

I know @zakird suggested some package reorganization in his review, too.

ajholland commented:
@cdzombak 

I'll make sure to keep the readme/docs up to date once we get this stuff ironed out. 

In Zgrab we have two massive structs, Config and GrabData. In order to add a new protocol a user has to define a config struct and add it to Config and then define a separate log/grab struct and add that to GrabData. 

By making it one struct we reduce the number of structs the user has to define and the different places it has to be defined. 

I can see where that would be confusing and an unneeded simplification, I'll change that. 

+1 On the package reorganization

cdzombak commented on main.go:
```diff
@@ -0,0 +1,23 @@
+		}
+	}
+
+	c := zlib.MakeNewController()
```
In line with https://github.com/zmap/zgrab2/pull/1#discussion_r1288295481 and the `MakeHandler` function elsewhere, this function should just be called `MakeController` — let's be consistent with naming conventions (in this case, `MakeThing` instead of `MakeNewThing`).

ajholland replied:
Main package named zgrab or zgrab2?

dadrian replied:
zgrab2 because that's the git repository. You can import it as zgrab if you want. We'll fix this all up once zgrab1 is dead.

cdzombak commented:
@ajholland as always, happy to discuss any feedback in person, too. I'll be at BBB tomorrow.

ajholland commented:
If any of you have time I'd appreciate a review again. 

Current status: 
Added input & dynamically parsing cidr/ip/domain
Added zflags & modifications to flags library, currently hosted at github.com/ajholland/zflags
Added ability to add module to zproto without editing a zgrab file
Added lookup

Need to add:
- [ ] refactor names, i.e. zproto? BaseProtocol or BaseModule? protocolResult/protocolResponse, **xssh/ssh?**
- [ ] dns lookup of domain
- [ ] conn abstraction/connect to ip
- [ ] finish implementing ssh, http, tls banner gets

zakird commented:
Let's schedule time on Tuesday to go over this.

On Fri, Aug 4, 2017 at 4:40 PM, Alex Holland <notifications@github.com>
wrote:

> If any of you have time I'd appreciate a review again.
>
> Current status:
> Added input & dynamically parsing cidr/ip/domain
> Added zflags & modifications to flags library, currently hosted at
> github.com/ajholland/zflags
> Added ability to add module to zproto without editing a zgrab file
> Added lookup
>
> Need to add:
>
>    - refactor names, i.e. zproto? BaseProtocol or BaseModule?
>    protocolResult/protocolResponse, *xssh/ssh?*
>    - dns lookup of domain
>    - conn abstraction/connect to ip
>    - finish implementing ssh, http, tls banner gets
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zmap/zgrab2/pull/1#issuecomment-320348213>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMSUC9lugBLxZpEj-8Jf70GOGM4esHZks5sU4GngaJpZM4OfrXR>
> .
>

zakird commented on README.md:
````diff
@@ -1,4 +1,85 @@
+```
+`Application Options` must be the initial section name. Other section names should correspond exactly to the relevant zgrab2 module. The default name for each module is the command name. If the same module is to be used multiple times then `name` must be specified and unique. 
+
+## Contributing
````
I'd probably change this to "Adding New Protocols"

zakird commented on main/main.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+func main() {
+	if _, err := zgrab2.ParseFlags(); err != nil {
```
I'd probably name the package `zgrab` and not `zgrab2`.

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+	for i := 0; i < workers; i++ {
+		go func() {
+			for obj := range processQueue {
+				//divide up, run, consolidate
```
remove comment

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	// Read the input, send to workers
+	input := bufio.NewReader(config.inputFile)
```
Probably should be an argument to `Process`.

ajholland replied:
pass in as reader

ajholland replied:
`process` is in the same package as `config`. Should they both just be removed from the arguments list and used in this form? I don't see the point of calling them from main (i.e. zgrab2.Process(zgrab2.config.inputFile))

ajholland commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+	// Read the input, send to workers
+	input := bufio.NewReader(config.inputFile)
+	for {
+		obj, err := input.ReadBytes('\n')
```
support different inputs: 
ip
domain
cidr block
ip, domain <- not implemented yet

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+		} else if err != nil {
+			log.Error(err)
+		}
+		st := string(obj)
```
Check that this doesn't break something unicode.

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+			log.Error(err)
+		} else {
+			if domain == "" {
+				for _, ip := range ipnet {
```
need to handle ip,domain case.

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+		ipnet, domain, err := ParseInput(st[:len(st)-1]) //remove newline
+		if err != nil {
+			log.Error(err)
+		} else {
```
probably shouldn't be an else

zakird commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,124 @@
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			log.Error(err)
```
put a continue

zakird commented on zgrab2/utility.go:
```diff
@@ -0,0 +1,63 @@
+	}
+}
+
+func inc(ip net.IP) {
```
name this something useful sounding :p

zakird commented on zgrab2/utility.go:
```diff
@@ -0,0 +1,63 @@
+	}
+}
+
+func dupIP(ip net.IP) net.IP {
```
give this a better name :|

zakird commented on zmodules/http.go:
```diff
@@ -0,0 +1,73 @@
+// Per module initialization call
+func init() {
+	var httpConfig HTTPConfig
+	cmd, err := zgrab2.AddCommand("http", "HTTP Banner Grab", "Grab a banner over HTTP", &httpConfig)
```
I might name this something like register scanner or the like

zakird commented on zmodules/http.go:
```diff
@@ -0,0 +1,73 @@
+func (x *HTTPConfig) Validate(args []string) error {
+	zgrab2.ValidateHighLevel()
+
+	zgrab2.RegisterLookup(x.Name, *x)
```
Figure out how to pull these two commands out o the module and into the framework.

zakird commented on zmodules/http.go:
```diff
@@ -0,0 +1,73 @@
+	return nil
+}
+
+func (x HTTPConfig) GetBanner() (interface{}, error) {
```
I might name this something more generic like `Scan()` since not all protocols have banners.

zakird commented on zmodules/http.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/zmap/zgrab2/zgrab2"
+)
+
+type HTTPConfig struct {
```
I might name this "HTTPScanner" or "HTTPModule".

zakird commented on zmodules/ssh.go:
```diff
@@ -0,0 +1,40 @@
+	"github.com/zmap/zgrab2/zgrab2"
+)
+
+type SSHConfig struct {
```
Get opinion from @dadrian on segmenting flags out into a sub-struct.

zakird commented on zmodules/ssh.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// per module per routine initialization call
+func (x SSHConfig) Initialize() {
```
might just name this `PerRoutineInitialize`.

zakird commented:
I could see either way. @dadrian thoughts?

On Wed, Aug 9, 2017 at 10:14 AM Alex Holland <notifications@github.com>
wrote:

> *@ajholland* commented on this pull request.
> ------------------------------
>
> In zgrab2/processing.go
> <https://github.com/zmap/zgrab2/pull/1#discussion_r132247688>:
>
> > +		outputDone.Done()
> +	}()
> +	//Start all the workers
> +	for i := 0; i < workers; i++ {
> +		go func() {
> +			for obj := range processQueue {
> +				//divide up, run, consolidate
> +				result := RunGrabWorker(obj, mon)
> +				outputQueue <- result
> +			}
> +			workerDone.Done()
> +		}()
> +	}
> +
> +	// Read the input, send to workers
> +	input := bufio.NewReader(config.inputFile)
>
> process is in the same package as config. Should they both just be
> removed from the arguments list and used in this form? I don't see the
> point of calling them from main (i.e.
> zgrab2.Process(zgrab2.config.inputFile))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zmap/zgrab2/pull/1#discussion_r132247688>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMSUIXGIgZlp8J7WdMHweDB401ETTPcks5sWekFgaJpZM4OfrXR>
> .
>

dadrian commented on zgrab2/handler.go:
```diff
@@ -0,0 +1,17 @@
+		mon.statuses <- status_failure
+		err = &e
+	}
+	resp := moduleResponse{Result: res, Error: err}
```
I'm confused. Interfaces are already effectively pointers, why not just pass the error directly? There's no situation where resp.Error is nil and e isn't also nil.

ajholland replied:
Similar case to strings where omitempty wouldn't actually omit an empty err since it would be "null" instead of nil.

dadrian commented on zgrab2/interface.go:
```diff
@@ -0,0 +1,50 @@
+	cmd.FindOptionByLongName("name").Default = []string{name}
+}
+
+var lookups map[string]Module
```
This should be a map of string to pointer to module, as discussed offline.

dadrian commented on zgrab2/interface.go:
```diff
@@ -0,0 +1,50 @@
+	lookups[name] = m
+}
+
+func NumActions() uint {
```
This may as well be an int since `len` returns an int.

dadrian commented on zgrab2/multiple.go:
```diff
@@ -0,0 +1,31 @@
+	"github.com/ajholland/zflags"
+)
+
+type MultipleConfig struct {
```
Not convinced this struct is named well. A config file doesn't necessarily define multiple scans, does it? Why isn't this part of the other the regular config struct? Can the config file override the global config things like `GOMAXPROCS` or input/output file?

ajholland replied:
I thought we predetermined that a config file **does** necessarily define multiple scans?

And yes, config file override's the global config things. The use cases should be single scan: all options entered through cli. Multiple scan: all options entered through config file, but also supports global config things being entered from cli, however if you specify both (I don't see why) config file takes precedence.

dadrian commented on zgrab2/multiple.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+// Validates the options sent to MultipleConfig, and parses the configFile
+func (x *MultipleConfig) Validate(args []string) error {
```
This config stuff is getting really hard to keep track of. What's the rationale against having a single config struct, a single `ConfigFileName` string, and then just reading the file at the start if passed?

ajholland replied:
The rational would be that someone could specify `./zgrab2 --config-file="mult.ini" http --port=8080`, which would make no sense. Either do a multiple scan, or do a single http. 

I think the naming here is probably the main issue, maybe instead of MultipleConfig it should be MultipleModules?

dadrian commented on zgrab2/multiple.go:
```diff
@@ -0,0 +1,31 @@
+
+// Validates the options sent to MultipleConfig, and parses the configFile
+func (x *MultipleConfig) Validate(args []string) error {
+	if x.ConfigFileName == config.InputFileName {
```
This seems like an excessive error check. Won't reading input just fail in this case?

dadrian commented on zgrab2/processing.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+type Grab struct {
+	IP     *string                   `json:"ip,omitempty"`
```
Why is this a pointer? It's exceedingly easy to have concurrency bugs with pointers.

dadrian replied:
(I'm assuming you send this through a channel at some point)

ajholland replied:
omitempty for json marshaling only works if the type is set to nil. A string can't be set to nil and the empty string isn't empty according to omitempty. 
It does get sent through a channel

ajholland commented on zgrab2/config.go:
```diff
@@ -0,0 +1,82 @@
+
+	// Validate Go Runtime config
+	if config.GOMAXPROCS < 0 {
+		log.Fatal("invalid GOMAXPROCS (must be at least 1, given %d)", config.GOMAXPROCS)
```
https://github.com/zmap/zgrab/blame/4c4f01ee32a5b6340c1d3b8f5300502549672e5d/main.go#L140

Past David says it does :p

ajholland commented on zgrab2/config.go:
```diff
@@ -0,0 +1,82 @@
+		if config.logFile, err = os.Create(config.LogFileName); err != nil {
+			log.Fatal(err)
+		}
+		log.SetOutput(config.logFile)
```
yup, default is stderr

ajholland commented on zgrab2/config.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func init() {
+	config.Multiple.ContinueOnError = true //set default for multiple value
```
When running Multiple modules someone may only want to run a module if the preceding modules for that ip/domain succeeded.

ajholland commented on main/main.go:
```diff
@@ -0,0 +1,30 @@
+
+	"github.com/ajholland/zflags"
+	log "github.com/sirupsen/logrus"
+	"github.com/zmap/zgrab2/zgrab2"
```
🤷‍♂️  I did it for directory cleanliness but I can bump it all up

ajholland commented on README.md:
````diff
@@ -1,4 +1,85 @@
+}
+```
+
+The `Validate()` function should first make a call to `zgrab2.ValidateHighLevel()` and lastly call `zgrab2.RegisterLookup()`.
````
I agree. I got rid of requiring ValidateHighLevel(), the issue is still the RegisterLookup function and corresponding call. Currently, the flag library parser goes through the .ini file, when it reaches the end of a section it calls Validate() on that section. I register the lookup at the end of that Validate so that the config doesn't get overwritten by a following config. The reason the config would be overwritten is because the parser requires pointers to structs. (Line 74 above, has to do with the way they use reflect package, goroutines panic otherwise)
I'm still stuck on two related issues here. A) how to remove the RegisterLookup call, and B) how to change the interfaces to work on pointer receivers.
Suggestions? We discussed earlier about changing the flag library but at this point that would involve refactoring a large part of the library or using the reflect package to deep copy the data from one pointer to a new pointer.

ajholland commented on interface.go:
```diff
@@ -0,0 +1,51 @@
+
+func RegisterLookup(name string, m Module) {
+	if lookups == nil {
+		lookups = make(map[string]*Module, 10)
```
Do this in init() and remove hard coded 10

ajholland commented on interface.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+type Module interface {
+	Scan(ip net.IP) (interface{}, error)
```
In depth comments

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+	Domain string
+}
+
+type moduleResponse struct {
```
public

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+// GrabWorker calls handler for each action
+func RunGrabWorker(input grabTarget, m *Monitor) []byte {
```
private

ajholland replied:
pointer grabTarget

ajholland replied:
rename grabTarget to target and this function should be grabTarget

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+type Grab struct {
+	IP     *string                   `json:"ip,omitempty"`
```
shouldn't be pointer

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+func RunGrabWorker(input grabTarget, m *Monitor) []byte {
+	moduleResult := make(map[string]moduleResponse)
+
+	for _, action := range lookups {
```
keep second list of names because map order undefined

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+		}
+	}
+
+	var ipstr *string
```
delet

ajholland commented on processing.go:
```diff
@@ -0,0 +1,127 @@
+	a := Grab{IP: ipstr, Domain: input.Domain, Data: moduleResult}
+	result, err := json.Marshal(a)
+	if err != nil {
+		log.Fatal(err)
```
fatalf

ajholland commented on processing.go:
```diff
@@ -0,0 +1,128 @@
+			if _, err := out.Write(result); err != nil {
+				log.Fatal(err)
+			}
+			if _, err := out.Write([]byte("\n")); err != nil {
```
WriteByte

ajholland commented on processing.go:
```diff
@@ -0,0 +1,128 @@
+				log.Fatal(err)
+			}
+		}
+		outputDone.Done()
```
defer

ajholland commented on processing.go:
```diff
@@ -0,0 +1,128 @@
+		}
+
+		if ipnet != nil {
+			for _, ip := range ipnet {
```
fix this to not waste a ton of memory, actual net.ipnet

ajholland commented on handler.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+// runHandler will call perRoutineInitialize, Scan, and respond with a protocol response, data unmarshalled, to the worker
+func runHandler(module Module, mon *Monitor, ip net.IP) (string, ModuleResponse) {
```
rethink function name

ajholland replied:
RunModule, public

ajholland replied:
rename interface.go to module.go move this to there

ajholland commented on interface.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/ajholland/zflags"
+)
+
+type Module interface {
```
ScanModule

ajholland commented on utility.go:
```diff
@@ -0,0 +1,86 @@
+		if ip := net.ParseIP(s); ip != nil {
+			return []net.IP{ip}, "", nil
+		} else {
+			ips, err := net.LookupIP(s)
```
@dadrian Currently I parse a cidr block and put it in a list. Bad idea. Trying to remove that and put it as a IPNet instead. I'm blocking on this line here, where I lookup the ip's associated with a domain. Should I just return 4 types, ([]net.IP, string, net.IPNet, error) or only perform a scan on the first IP associated with a domain? Presently zgrab2 will lookup a domain, find all IP's associated with that domain, and then scan all of them.

dadrian replied:
Only scan the first one for now.
